### PR TITLE
[SPARK-14489][SPARK-14153][ML][PYSPARK] Support dropping NaN predicted values in RegressionEvaluator

### PR DESCRIPTION
**Pull request by MLnick**
As discussed in [SPARK-14489](https://issues.apache.org/jira/browse/SPARK-14489), when using `ALSModel` to predict on a test set, the model returns `NaN` when the user/item is in the test set but not the training set, since the model has not computed factor(s) for that user and/or item. 

This PR adds support to `RegressionEvaluator` to drop rows where the value of `predictionCol` is `NaN`. This should not be used in the general case (since a bad regression model may produce a lot of `NaN`s and one would not want to ignore those but rather fix the underlying issue), but allows ALS to be used in cross-validation settings even when this situation occurs (which may be quite common on larger, sparser datasets). Thus it is an `expertParam` and the default is `false`. 
## How was this patch tested?

New unit tests in `RegressionEvaluatorSuite` and doc string tests in `evaluation.py`.

cc @srowen @sethah @jkbradley

### CODE REVIEW

**Comment by holdenk** [comment]
It sounds like people are running into this often when using cross-validation - would it make sense to also mention this in the k-fold docstring or examples? (Just a minor suggestion not to block or anything).

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala [hunk 1]
```diff
@@ -86,8 +106,9 @@ final class RegressionEvaluator @Since("1.4.0") (@Since("1.4.0") override val ui
 
     val predictionAndLabels = dataset
       .select(col($(predictionCol)).cast(DoubleType), col($(labelCol)).cast(DoubleType))
-      .rdd.
-      map { case Row(prediction: Double, label: Double) =>
+      .na.drop("any", if ($(dropNaN)) Seq($(predictionCol)) else Seq())
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala]: This also drops `null` values. I'm not sure how likely this is to happen, but the documentation should probably note it drops `NaN` _and_ `null` values.

File: python/pyspark/ml/evaluation.py [hunk 2]
```diff
@@ -187,6 +187,13 @@ class RegressionEvaluator(JavaEvaluator, HasLabelCol, HasPredictionCol):
     0.993...
     >>> evaluator.evaluate(dataset, {evaluator.metricName: "mae"})
     2.649...
+    >>> scoreAndLabels = [(4.0, 5.0), (4.0, 1.0), (float('nan'), 2.0),
+    ...   (1.0, 3.0), (float('nan'), 4.0)]
+    >>> dataset = sqlContext.createDataFrame(scoreAndLabels, ["raw", "label"])
+    ...
+    >>> evaluator = RegressionEvaluator(predictionCol="raw").setDropNaN(True)
+    >>> evaluator.evaluate(dataset)
```
Comment (sethah) [inline python/pyspark/ml/evaluation.py]: This doesn't quite mirror the Scala test, since the scala test first checks that the result is `NaN` when `dropNaN` is true. Shall we do the same check here?

File: python/pyspark/ml/evaluation.py [hunk 3]
```diff
@@ -197,18 +204,24 @@ class RegressionEvaluator(JavaEvaluator, HasLabelCol, HasPredictionCol):
                        "metric name in evaluation (mse|rmse|r2|mae)",
                        typeConverter=TypeConverters.toString)
 
+    dropNaN = Param(Params._dummy(), "dropNaN",
+                    "whether to drop rows where 'predictionCol' is NaN. NOTE - only set this to " +
+                    "True if you are certain that NaN predictions should be ignored! " +
+                    "(default: False)",
+                    typeConverter=TypeConverters.toBoolean)
```
Comment (sethah) [inline python/pyspark/ml/evaluation.py]: nit: move typeConverter up a line.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala [hunk 4]
```diff
@@ -69,7 +69,27 @@ final class RegressionEvaluator @Since("1.4.0") (@Since("1.4.0") override val ui
...
-  setDefault(metricName -> "rmse")
+  /**
+   * Param for whether to drop rows where 'predictionCol' is NaN. NOTE - only set this to
+   * true if you are certain that NaN predictions should be ignored!
+   * (default: false)
+   *
+   * @group expertParam
+   */
+  @Since("2.0.0")
+  val dropNaN: BooleanParam = new BooleanParam(this, "dropNaN",
```
Comment (sethah) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala]: Making this a Boolean parameter called `dropNaN` makes it less extensible in the future if we wish to implement more than just one possible NaN behavior. If we don't envision adding any other behavior then I guess this is good, but otherwise we could make a String param and limit its options to drop or raise an error for now.

**Comment by sethah** [comment]
From a high level, one concern is that this seems to be a bit of band-aid fix. If the only scenario where this is a problem is using ALS in cross validation then it would seem better to address the problem at its root. I am trying to think of other scenarios where a predictor would output `NaN` and whether this behavior is desirable in that case. In the [Jira](https://issues.apache.org/jira/browse/SPARK-14489), @jkbradley mentioned linear models with too little regularization. Are there any other scenarios that come to mind? I'll give it some thought.

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala (hunk 1 above)
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala]: Good point, will add `null` to test cases. I don't think it's likely in practice. But actually if nulls do exist in the dataset, it's worse than NaN from a correctness point of view, as either a NPE will be thrown, or it will be treated as 0 => 0 squared error for that datapoint, but the denominator will still be added for the mean calculation. So MSE will be biased low.

File: python/pyspark/ml/evaluation.py (hunk 2 above)
Comment (MLnick) [inline python/pyspark/ml/evaluation.py]: will do

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala (hunk 4 above)
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala]: Let me see how nulls are handled given nullable input columns, and perhaps the possible strategies can be adjusted.

**Comment by MLnick** [comment]
@holdenk yes I think it makes sense to add something to docs on cross-val to illustrate use cases.

**Comment by MLnick** [comment]
@sethah you're correct, this is a a bit of a band-aid fix. However, the real fix is getting CrossValidator to handle cases like this in a principled and generic way (and/or to change the behavior of ALS in predicting missing users). But even fixing ALS for missing users, I think the issue will still arise for missing items.

As for the "ideal" fix in CrossValidator, it seems to be from your JIRA comment that this will be fairly complex. So until we can fix that, users can not use ALS in cross-validation in many cases.

I've kept it an expertParam and tried to highlight one should only use it when you know what you're doing. I think we could also deprecate this option once the "real" fix comes in CrossValidator...

@srowen @jkbradley ?

File: mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala (hunk 4 above)
Comment (MLnick) [inline mllib/src/main/scala/org/apache/spark/ml/evaluation/RegressionEvaluator.scala]: Currently if `nulls` are present `RegressionEvaluator` throws a `MatchError`. I think we should either (a) disallow nullable columns explicitly with a schema check in `evaluate` - this can then provide a more understandable error message too; (b) allow nulls, but ignore them for both prediction and label col.

I think nulls in the input for this case are unlikely and probably a result of bad data or user error somewhere along the line. So I'd prefer option (a). This then means the `dropNaN` setting will only apply to NaNs.

**Comment by sethah** [comment]
@MLnick Good points. In my mind there are two scenarios here:

1. This is a quick fix for users wanting to do cross validation with ALS/recommenders in Spark, until we can fully address the issue in the ALS algorithm.
2. This is an improvement to Evaluators in general, giving users flexibility to ignore NaNs in whatever scenario produces them.

If (1) is true then, as you said, we can think about deprecating this in the future since it may happen that we can think of no specific use case for it once (if?) ALS stops predicting NaNs on new data. If (2) is true,  perhaps we should consider adding this to _all_ evaluators? Again, I'd be interested to hear other use cases. One I thought of is a Naive Bayes classifier with no smoothing predicting on unseen words in text classification, but I wasn't able to produce a similar failure in the bit of time I spent on it. Either way, I think this is an improvement, but just wanted to be a bit more explicit on the _why_ and how it might affect scope.

**Comment by MLnick** [comment]
@sethah @holdenk @jkbradley I thought about this some more. I can't realistically think of a scenario apart from the ALS one where handling NaNs in the evaluator is desirable.

So actually I think this should rather go into ALS itself - I'll call the param something like `unknownUserItemStrategy`. The default can be to return NaN as it does currently. We can make an option (perhaps called "skip" or "filter") that filters out NaNs in the prediction DF. This would allow this option to be used in cross-validation. This will also make it extensible for future potential additions such as using the "average user"  factor, or whatever other strategy.

**Comment by MLnick** [comment]
Opened #12896
